homeassistant/components/lamarzocco/coordinator.py

```diff
@@ -149,14 +97,44 @@ async def _async_handle_request(
             _LOGGER.debug(ex, exc_info=True)
             raise UpdateFailed("Querying API failed. Error: %s" % ex) from ex
 
-    def async_get_ble_device(self) -> BLEDevice | None:
-        """Get a Bleak Client for the machine."""
-        # according to HA best practices, we should not reuse the same client
-        # get a new BLE device from hass and init a new Bleak Client with it
-        if not self._use_bluetooth:
-            return None
 
-        return async_ble_device_from_address(
-            self.hass,
-            self.lm.lm_bluetooth.address,
-        )
+class LaMarzoccoMachineUpdateCoordinator(
+    LaMarzoccoUpdateCoordinator[LaMarzoccoMachine]
+):
+    """Class to handle fetching data from the La Marzocco API for a machine."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        cloud_client: LaMarzoccoCloudClient,
+        local_client: LaMarzoccoLocalClient | None,
+        bluetooth_client: LaMarzoccoBluetoothClient | None,
+    ) -> None:
+        """Initialize coordinator."""
+
+        super().__init__(hass, cloud_client, local_client, bluetooth_client)
+
+        if local_client is not None:
+            _LOGGER.debug("Init WebSocket in background task")
+
+            self.config_entry.async_create_background_task(
+                hass=self.hass,
+                target=self.device.websocket_connect(
+                    notify_callback=lambda: self.async_set_updated_data(None)
+                ),
+                name="lm_websocket_task",
+            )
+
+            async def websocket_close(_: Any | None = None) -> None:
+                if local_client.websocket is not None and local_client.websocket.open:
+                    local_client.terminating = True
+                    await local_client.websocket.close()
+
+            self.config_entry.async_on_unload(
+                hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, websocket_close)
+            )
+            self.config_entry.async_on_unload(websocket_close)
```

Comment:
I am not sure if we always do something like this in the constructor. I can imagine some reasons against it, but I also don't have any information if we do this in more places. WDYT

Reply:
My main issue is that there is no way of knowing during the coordinator update if it's the first refresh (maybe it would make sense to add an empty method to `async_first_refresh` which implementations can fill with life), because at it is atm I don't like checking for a new var is_first_refresh or something during an update

Reply:
You could also just make a function like `async_setup` and then call that before running the `async_first_refresh`. But the thing also is, I am now saying this, but I have no data on if this is a common thing for coordinators to have

Reply:
I mean I could also overwrite `async_config_entry_first_refresh` to first do the setup and only then do `super().async_config_entry_first_refresh`. I don't know either, for a clean solution I'd think an overwritable function inside the coordinator would make sense, for proper & consitant exception handling